python/cell_reader.py

```diff
@@ -195,6 +195,80 @@ def get_factors(metadata):
     return factors
 
 
+def get_cell_sets(clusters, lookup):
+    '''
+    >>> from collections import namedtuple
+    >>> Cluster = namedtuple('Cluster', ['name', 'cell_ids'])
+    >>> clusters = {
+    ...   1: Cluster('pyramidal L4', ['104', '110', '111']),
+    ...   3: Cluster('vascular smooth muscle', ['1', '2', '3'])
+    ... }
+    >>> lookup = {
+    ...   'vascular smooth muscle': 'vasculature',
+    ...   'pyramidal L4': 'excitatory neurons'
+    ... }
+    >>> cell_sets = get_cell_sets(clusters, lookup)
+    >>> list(cell_sets.keys())
+    ['version', 'datatype', 'tree']
+    >>> cell_sets['datatype']
+    'cell'
+    >>> list(cell_sets['tree'][0].keys())
+    ['name', 'children']
+    >>> cell_sets['tree'][0]['name']
+    'Cell Type Annotations'
+    >>> sorted([ n['name'] for n in cell_sets['tree'][0]['children'] ])
+    ['excitatory neurons', 'vasculature']
+    '''
+
+    # The parameter `lookup` is a dict mapping
+    # subclusters to clusters: `{ Subcluster Name: Cluster Name }`
+    # This `lookup` mapping can be used to fill in an intermediate
+    # dict `hierarchy`, closer to the data structure we want to output.
+    # ```
+    # {
+    #   Cluster A: {
+    #     Subcluster A: [1, 2],
+    #     Subcluster B: [3, 4]
+    #   },
+    #   Cluster B: {...}
+    # }
+    # ```
+    hierarchy = {cluster_name: {} for cluster_name in lookup.values()}
+    for c in clusters.values():
+        subcluster = {c.name: c.cell_ids}
+        cluster_name = lookup.get(c.name)
+        cluster_dict = hierarchy.get(cluster_name)
+        cluster_dict.update(subcluster)
+
+    # Use the `hierarchy` dict to fill in an object
+    # conforming to the `cell_sets.json` schema.
+    cluster_nodes = []
+    for cluster_name in sorted(hierarchy.keys()):
+        cluster_dict = hierarchy[cluster_name]
+        subcluster_nodes = []
+        for subcluster_name in sorted(cluster_dict.keys()):
+            subcluster = cluster_dict[subcluster_name]
+            subcluster_nodes.append({
+                'name': subcluster_name,
+                'set': subcluster
+            })
+        cluster_nodes.append({
```

Comment:
consider a helper function and array comprehensions when things get a little deep.

Reply:
I got previous feedback that the array comprehension was more difficult to understand. Ultimately I think the tests are helpful here to ensure that the code does what it is supposed to do, regardless of implementation https://github.com/hubmapconsortium/vitessce-data/pull/94#discussion_r426164341